core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -51,7 +52,8 @@
  */
 @ThreadSafe
 public abstract class AbstractClient implements Client {
-  private static final Logger LOG = LoggerFactory.getLogger(AbstractClient.class);
+  private static final Logger LOG =
```

Comment:
This will impact all clients, not just metrics

Reply:
I am aware

Reply:
In particular, if metrics master client fails to connect to master and logs some warnings, other clients, e.g. file system master client, job master client, etc, will suppress their logs for 10 seconds. So users may only see metrics master client log messages, even though other clients are failing to connect as well.

In general, I don't think we want to use sampling loggers at a static scope because it makes it easy to suppress log messages that we actually want.

Reply:
I disagree. I feel that the amount of logging that the AbstractClient does is very high. I understand that using a SamplingLogger will cause contention of log messages by different clients but I feel that the amount of logging that is produced from this class is too high. Take for example the following outputs before and after attempting to open a file over a 10-second period.

```log
2019-04-30 12:08:02,933 WARN  alluxio.AbstractClient (SamplingLogger.java:warn) - Failed to connect (1) with MetricsMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:08:13,112 WARN  alluxio.AbstractClient (SamplingLogger.java:warn) - Failed to connect (13) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
```

Without the `SamplingLogger` I get 

```log
2019-04-30 12:05:22,454 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (1) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:22,520 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (2) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:22,637 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (3) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:22,848 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (4) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:23,280 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (5) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:24,114 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (6) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:25,519 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (1) with MetricsMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:25,867 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (7) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:28,525 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (1) with MetricsMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:28,910 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (8) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:31,534 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (1) with MetricsMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:32,132 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (9) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:34,542 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (1) with MetricsMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:35,362 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (10) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:37,549 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (1) with MetricsMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:38,604 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (11) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:40,557 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (1) with MetricsMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
2019-04-30 12:05:41,792 WARN  alluxio.AbstractClient (AbstractClient.java:connect) - Failed to connect (12) with FileSystemMasterClient @ zac-alluxio/10.0.0.8:19998: Failed to handshake with master zac-alluxio/10.0.0.8:19998 to load cluster default configuration values: UNAVAILABLE: io exception
```

Most of this is just the same message for multiple different clients for the same reason. Do you feel that this kind of output is still useful?

Reply:
I'm totally in agreement that the logging is too much.

For the example you give, I see only two clients, metrics heartbeat and filesystem master. In this situation, I'd want to see two warnings, one for metrics and one for filesystem master. Do you think one warnings per client is too verbose?

Reply:
I think for the number of separate clients that talk to the master within the JVM, that together and with retries they perform too much logging. Preferably each type of client would have more fine-grained control over when they log messages. Unfortunately I don't think we'll easily be able get that kind of granularity because we don't propagate any errors from AbstractClient back to FileSystemMasterClient, MetricsMasterClient, etc..

What I've seen is that if a client fails to connect, usually all the other clients fail with the exact same reason. I can't recall ever seeing a case where clients failed for different reasons at the same time. I was aware that adding the SamplingLogger would cause us to lose logs, but due to the liberal nature that we log within the AbstractClient I felt it warranted the use. Maybe you've seen other cases where errors are different- but I haven't

Otherwise we could possibly move the `WARN` levels to `DEBUG`, and then `ERROR` or `WARN` when the retryPolicy reaches certain points (multiples of 15 or 20 maybe?)

Reply:
How about we put all `AbstractClient` retry logging at debug level? If it continues to fail, we will eventually throw an exception, which should be handled or logged by whatever called the RPC.

If we **really** want to put non-debug error logging in `AbstractClient`, we could put it after the final retry fails, right before throwing an exception. But I think this is unnecessary since the place that catches the exception will log it anyway.

Reply:
I have moved everything to `DEBUG` logging

Reply:
Will using the sampling logger lead to confusing output since we log things like the attempt count?

Reply:
No. My bad, I missed removing the sampling logger. Will be fixed soon